### PR TITLE
Ignore Path MTU Discovery on UDP server socket

### PR DESCRIPTION
**Pull request by rgacogne**
### Short description

It might help prevent Path MTU poisoning attacks.
Untested, don't merge it yet!

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [ ] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: pdns/iputils.cc [hunk 1]
```diff
@@ -136,6 +136,25 @@ int SSetsockopt(int sockfd, int level, int opname, int value)
...
+    SSetsockopt(sockfd, IPPROTO_IP, IP_MTU_DISCOVER, IP_PMTUDISC_OMIT);
+    return;
+  }
+  catch(const std::exception& e) {
+    /* failed, let's try IP_PMTUDISC_DONT instead */
+  }
+#endif /* IP_PMTUDISC_OMIT */
+
+  /* IP_PMTUDISC_DONT disables Path MTU discovery */
+  SSetsockopt(sockfd, IPPROTO_IP, IP_MTU_DISCOVER, IP_PMTUDISC_DONT);
```
Comment (rgacogne) [inline pdns/iputils.cc]: We should check that `IP_MTUDISC_DONT` is defined. Note that unbound set the don't fragment flag on systems that don't have `IP_MTUDISC_DONT`: https://github.com/NLnetLabs/unbound/blob/master/services/listen_dnsport.c#L571

**Comment by rgacogne** [comment]
As a datapoint, Knot decided against doing this: https://gitlab.labs.nic.cz/knot/knot-dns/issues/467

File: pdns/iputils.cc (hunk 1 above)
Comment (rgacogne) [reply pdns/iputils.cc]: Bind is doing the same: https://gitlab.isc.org/isc-projects/bind9/blob/master/lib/isc/unix/socket.c#L2396

**Comment by mnordhoff** [comment]
Knot decided against it in 2016 when fragmentation attacks weren't in vogue, though.

File: pdns/iputils.cc (hunk 1 above)
Comment (rgacogne) [reply pdns/iputils.cc]: Added the missing check for `IP_MTUDISC_DONT`. I'm still undecided about setting DF=1.

**Comment by rgacogne** [comment]
Rebased to fix a conflict.

**Comment by rgacogne** [comment]
As pointed out by @hdais `NSD` did bite the bullet as well recently: https://www.nlnetlabs.nl/bugs-script/show_bug.cgi?id=4235

**Comment by rgacogne** [comment]
Move this PR into the rec-4.2.0-beta1 milestone. Now is the time to speak if you are against this change ;-)

**Comment by jedisct1** [comment]
Hi Remi,

That change seems to be what rmade DNSCrypt servers running `dnsdist` unreliable after they upgraded to version 1.4.0.

As you know since you implemented the protocol, DNSCrypt requires questions sent over UDP to be as large as responses, using padding.

Blocking fragmented questions prevents large responses from being received.

Ignoring fragments in responses from authoritative servers fine, but even with unencrypted queries, I'm not sure that there is any value in dropping fragments on the server socket. A fragment attack would just allow the question to be modified, but a stub resolver would ignore a response for a different query.

**Comment by rgacogne** [comment]
Hi Frank,

Thanks a lot for reporting this! I have to admit I had completely forgotten that propriety of `DNSCrypt` when I submitted this change, sorry about that!
I'll re-evaluate whether it makes sense to keep setting `setSocketIgnorePMTU()` on our incoming sockets. At the very least I'll disable it for `DNSCrypt` binds.

**Comment by jedisct1** [comment]
Thank you so much Remi!

/cc @welwood08

**Pull request merged**